workflows/cloudify_system_workflows/snapshot.py

```diff
@@ -0,0 +1,574 @@
+            for node_instance_id, agent in node_instances.iteritems():
+                node_instance = client.node_instances.get(node_instance_id)
+                runtime_properties = node_instance.runtime_properties
+                old_agent = runtime_properties.get('cloudify_agent', {})
+                old_agent.update(agent)
+                runtime_properties['cloudify_agent'] = old_agent
+                client.node_instances.update(
+                    node_instance_id=node_instance_id,
+                    runtime_properties=runtime_properties)
+
+
+def _restore_agents_data(ctx, tempdir):
+    ctx.send_event('Updating cloudify agent data')
+    client = get_rest_client()
+    with open(os.path.join(tempdir, _AGENTS_FILE)) as agents_file:
+        agents = json.load(agents_file)
+    insert_agents_data(client, agents)
+
+
+def _restore_snapshot(ctx, config, tempdir, metadata):
+    # files/dirs copy
+    _copy_data(tempdir, config, to_archive=False)
+
+    # elasticsearch
+    es = _create_es_client(config)
+
+    _restore_elasticsearch(ctx, tempdir, es, metadata)
+
+    # influxdb
+    _restore_influxdb_3_3(ctx, tempdir)
+
+    # credentials
+    _restore_credentials_3_3(ctx, tempdir, config.file_server_root, es)
+
+    es.indices.flush()
+
+    # agents
+    _restore_agents_data(ctx, tempdir)
+    # end
+
+
+def _restore_snapshot_format_3_3(ctx, config, tempdir, metadata):
+    _restore_snapshot(ctx, config, tempdir, metadata)
+
+
+def _restore_snapshot_format_3_2(ctx, config, tempdir, metadata):
+    _restore_snapshot(ctx, config, tempdir, metadata)
+
+
+def recreate_deployments_environments(ctx):
+    ctx.load_deployments_contexts()
+    rest_client = get_rest_client()
+    for dep_id, dep_ctx in ctx.deployments_contexts.iteritems():
+        dep = rest_client.deployments.get(dep_id)
+        blueprint = rest_client.blueprints.get(dep_ctx.blueprint.id)
+        blueprint_plan = blueprint['plan']
+        tasks_graph = generate_create_dep_tasks_graph(
+            dep_ctx,
+            deployment_plugins_to_install=blueprint_plan[
+                'deployment_plugins_to_install'],
+            workflow_plugins_to_install=blueprint_plan[
+                'workflow_plugins_to_install'],
+            policy_configuration={
+                'policy_types': dep['policy_types'],
+                'policy_triggers': dep['policy_triggers'],
+                'groups': dep['groups']
+            }
+        )
+        try:
+            dep_ctx.internal.start_local_tasks_processing()
+            dep_ctx.internal.start_event_monitor()
+            tasks_graph.execute()
+        finally:
+            dep_ctx.internal.stop_local_tasks_processing()
+            dep_ctx.internal.stop_event_monitor()
+        ctx.send_event('Successfully created {0}\'s deployment '
+                       'environment'.format(dep_id))
+
+
+@workflow(system_wide=True)
+def restore(ctx, snapshot_id, recreate_deployments_envs, config, **kwargs):
```

Comment:
Perhaps it would be good to add some logs during the process of creating a snapshot or restoring one.